hazelcast/src/main/java/com/hazelcast/map/impl/operation/MapFetchIndexOperation.java

```diff
@@ -0,0 +1,328 @@
+    }
+
+    @Override
+    public Object getResponse() {
+        return response;
+    }
+
+    @Override
+    protected void readInternal(ObjectDataInput in) throws IOException {
+        super.readInternal(in);
+        indexName = in.readString();
+        partitionIdSet = in.readObject();
+        pointers = in.readObject();
+        sizeHint = in.readInt();
+    }
+
+    @Override
+    protected void writeInternal(ObjectDataOutput out) throws IOException {
+        super.writeInternal(out);
+        out.writeString(indexName);
+        out.writeObject(partitionIdSet);
+        out.writeObject(pointers);
+        out.writeInt(sizeHint);
+    }
+
+    @Override
+    public int getClassId() {
+        return MAP_FETCH_INDEX;
+    }
+
+    private boolean allIndexed(InternalIndex index, PartitionIdSet partitionIdSet) {
+        return partitionIdSet.stream().allMatch(index::hasPartitionIndexed);
+    }
+
+    protected int getMigrationTimestamp() {
+        return mapService.getMigrationStamp();
+    }
+
+    /**
+     *  Return only the {@code entries} from the partitions declared in {@code partitionIdSet}
+     *
+     * @param entries
+     * @param partitionIdSet
+     * @return list of entries belonging the given partitions
+     */
+    @Nonnull
+    private List<QueryableEntry> getOwnedEntries(
+            @Nonnull Collection<QueryableEntry> entries,
+            @Nonnull PartitionIdSet partitionIdSet
+    ) {
+        return entries.stream()
+                .filter(entry -> {
+                    int partitionId = partitionService.getPartitionId(entry.getKeyData());
+                    return partitionIdSet.contains(partitionId);
+                })
+                .collect(Collectors.toList());
+    }
+
+    private static boolean checkPointerForLookup(IndexIterationPointer pointer) {
+        return ((Comparable) pointer.getFrom()).compareTo(pointer.getTo()) == 0;
+    }
+
+    public static class MapFetchIndexOperationResult {
+        private final List<QueryableEntry> entries;
+        private final IndexIterationPointer[] pointers;
+
+        public MapFetchIndexOperationResult(
+                List<QueryableEntry> entries,
+                IndexIterationPointer[] pointers
+        ) {
+            this.entries = entries;
+            this.pointers = pointers;
+        }
+
+        public List<QueryableEntry> getEntries() {
+            return entries;
+        }
+
+        public IndexIterationPointer[] getPointers() {
+            return pointers;
+        }
+    }
+
+    public static final class MigrationDetectedException extends HazelcastException {
+        public MigrationDetectedException() {
+            super("Migration detected, operation is cancelled");
+        }
+    }
+
+    public static final class MissingPartitionException extends HazelcastException {
+        public MissingPartitionException() {
+            super("Some partitions are missing");
+        }
+    }
```

Comment:
Not sure about aim of two different exceptions. 
From logical point of view, it is correct, but de-facto it just detects that partition migration happens.

Reply:
If the reader does not differentiate between these 2 cases, we can merge them.